File: qing_voting_py/extractLayerFeat.py
import numpy as np
import math
import pickle
import scipy.io as sio
from FeatureExtractor import *
from config_voting import *

def extractLayerFeat(category_ls, set_type):
    extractor = FeatureExtractor(cache_folder=model_cache_folder, which_layer=VC['layer'], which_snapshot=25000)
    
    for category in category_ls:
        dir_img = Dataset['img_dir'].format(category)
        dir_anno = Dataset['anno_dir'].format(category)
        
        file_list = Dataset['{0}_list'.format(set_type)].format(category)
        assert(os.path.isfile(file_list))
        
        with open(file_list, 'r') as fh:
            content = fh.readlines()
        
        img_list = [x.strip().split() for x in content]
        img_num = len(img_list)
        print('total number of images for {1}: {0}'.format(img_num, category))
        
        feat_set = [None for nn in range(img_num)]
        for nn in range(img_num):
            file_img = os.path.join(dir_img, '{0}.JPEG'.format(img_list[nn][0]))
            assert(os.path.isfile(file_img))
            img = cv2.imread(file_img)
            # plt.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
            # plt.show()
            
            height, width, _ = img.shape
            
            file_anno = os.path.join(dir_anno, '{0}.mat'.format(img_list[nn][0]))
            assert(os.path.isfile(file_anno))
            mat_contents = sio.loadmat(file_anno)
            record = mat_contents['record']
            objects = record['objects']
            bbox = objects[0,0]['bbox'][0,int(img_list[nn][1])-1][0]
            bbox = [max(math.ceil(bbox[0]), 1), max(math.ceil(bbox[1]), 1), \
                    min(math.floor(bbox[2]), width), min(math.floor(bbox[3]), height)]
            patch = img[bbox[1]-1: bbox[3], bbox[0]-1: bbox[2], :]
            
            feat_set[nn] = extractor.extract_feature_image(patch)[0]
            
            if nn%100 == 0:
                print(nn, end=' ')
            
            
        print('\n')
        
        if not os.path.exists(Feat['cache_dir']):
            os.makedirs(Feat['cache_dir'])
            
        file_cache_feat = os.path.join(Feat['cache_dir'], '{0}_{1}_{2}.pickle'.format(category, dataset_suffix, set_type))
        with open(file_cache_feat, 'wb') as fh:
            pickle.dump(feat_set, fh)

File: qing_voting/extractLayerFeat.m
% extract deep network layer features

function extractLayerFeat(set_type, config_file)
try    
    eval(config_file);
catch
    keyboard;
end

fprintf('extract deep network layer features for "%s" set ...\n', set_type);

caffe.set_mode_gpu();
caffe.set_device(GPU_id);
net = caffe.Net(model, weights, 'test');

dir_img = sprintf(Dataset.img_dir, category);
dir_anno = sprintf(Dataset.anno_dir, category);

switch set_type
    case 'train'
        file_list = sprintf(Dataset.train_list, category);
    case 'test'
        file_list = sprintf(Dataset.test_list, category);
    otherwise
        error('Error: unknown set_type');
end        
assert(exist(file_list, 'file') > 0);
file_ids = fopen(file_list, 'r');
img_list = textscan(file_ids, '%s %d');
img_num = length(img_list{1});

select_num = img_num;
feat_set = cell(1, select_num);
for n = 1: select_num                       % for each image   
    
    file_img = sprintf('%s/%s.JPEG', dir_img, img_list{1}{n});
    img = imread(file_img);
    [height, width, ~] = size(img);
    if size(img, 3) == 1
        img = cat(3, img, img, img);
    end

    file_anno = sprintf('%s/%s.mat', dir_anno, img_list{1}{n});
    anno = load(file_anno);
    anno = anno.record;
    bbox = anno.objects(img_list{2}(n)).bbox;
    
    bbox = [max(ceil(bbox(1)), 1), max(ceil(bbox(2)), 1), min(floor(bbox(3)), width), min(floor(bbox(4)), height)];
    patch = img(bbox(2): bbox(4), bbox(1): bbox(3), :);
    scaled_patch = myresize(patch, caffe_dim, 'short');
    
    % compute deep network features
    data = single(scaled_patch(:,:,[3, 2, 1]));
    data = bsxfun(@minus, data, mean_pixel);
    data = permute(data, [2, 1, 3]);
    net.blobs('data').reshape([size(data), 1]);
    net.reshape();
    net.forward({data});
    layer_feature = permute(net.blobs(layer_name).get_data(), [2, 1, 3]);
    feat_set{n} = layer_feature;  
    
    if mod(n,100) == 0
        disp(n);
    end    
end % image

%%
MkdirIfMissing(Feat.cache_dir);
switch set_type
    case 'train'
        file_cache_feat = fullfile(Feat.cache_dir, sprintf('%s_%s_train.mat', category, dataset_suffix));
    case 'test'
        file_cache_feat = fullfile(Feat.cache_dir, sprintf('%s_%s_test.mat', category, dataset_suffix));
    otherwise
        error('Error: unknown type of dataset;\n');
end
save(file_cache_feat, 'feat_set', '-v7.3');

end % end of function

